python_modules/dagster/dagster/_model/decorator.py

```diff
@@ -0,0 +1,241 @@
+    """
+    field_set = {
+        **cls.__annotations__,
+        **({NT_CACHED_METHOD_CACHE_FIELD: Any} if enable_cached_method else {}),
+    }
+    base = NamedTuple(f"_{cls.__name__}", field_set.items())
+
+    if checked:
+        orig_new = base.__new__
+        checks_builder = LazyCheckBuilder(
+            cls.__annotations__,
+            EvalContext.capture_from_frame(2),  # this method, original decorator
+            1 if with_new else 0,
+        )
+
+        def __checked_new__(cls: TType, **kwargs):
+            for key, fn in checks_builder.checks.items():
+                fn(kwargs[key])
+
+            if enable_cached_method:
+                kwargs[NT_CACHED_METHOD_CACHE_FIELD] = {}
+
+            return orig_new(cls, **kwargs)
+
+        base.__new__ = __checked_new__  # type: ignore # unhappy with dropping positional args
+
+    if enable_cached_method:
+        base.__eq__ = __cached_method_safe_eq__
+        base.__hash__ = __cached_method_safe_hash__
+
+    return type(  # type: ignore
+        cls.__name__,
+        (cls, base),
+        {
+            "__iter__": _banned_iter,
+            "__getitem__": _banned_idx,
+            "__hidden_iter__": base.__iter__,
+            _MODEL_MARKER_FIELD: _MODEL_MARKER_VALUE,
+            "__annotations__": cls.__annotations__,
+        },
+    )
+
+
+@overload
+def dagster_model(
+    cls: TType,
+) -> TType: ...  # Overload for using decorator with no ().
+
+
+@overload
+def dagster_model(
+    *,
+    checked: bool = True,
+    enable_cached_method: bool = False,
+) -> Callable[[TType], TType]: ...  # Overload for using decorator used with args.
+
+
+@dataclass_transform(
+    kw_only_default=True,
+    frozen_default=True,
+)
+def dagster_model(
+    cls: Optional[TType] = None,
+    *,
+    checked: bool = True,
+    enable_cached_method: bool = False,
+) -> Union[TType, Callable[[TType], TType]]:
+    """A class decorator that will create an immutable model class based on the defined fields.
+
+    Args:
+        checked: Whether or not to generate runtime type checked construction.
+        enable_cached_method: Whether or not to support object instance level caching using @cached_method.
+    """
+    if cls:
+        return _namedtuple_model_transform(
+            cls,
+            checked=checked,
+            enable_cached_method=enable_cached_method,
+            with_new=False,
+        )
+    else:
+        return partial(
+            _namedtuple_model_transform,
+            checked=checked,
+            enable_cached_method=enable_cached_method,
+            with_new=False,
+        )
+
+
+@overload
+def dagster_model_with_new(
+    cls: TType,
+) -> TType: ...  # Overload for using decorator with no ().
+
+
+@overload
+def dagster_model_with_new(
+    *,
+    checked: bool = True,
+    enable_cached_method: bool = False,
+) -> Callable[[TType], TType]: ...  # Overload for using decorator used with args.
+
+
+def dagster_model_with_new(
```

Comment:
Ergonomically, this looks too specific of a variant. Do we really need first class support for the ability to override `__new__`?
Is it possible to achieve the same by instead:
1. using an independent constructor function where needed
2. overriding `__new__` in a subclass of a dagster_model decorated class

Reply:
context on this explained here https://github.com/dagster-io/dagster/pull/22043#discussion_r1626258879 

> Do we really need first class support for the ability to override __new__?

I think so - its a common pattern to coerce flexible inputs in to a canonical representation. This decorator sucks but I think having to go through a static constructor in callsites where you are trying to be flexible with inputs would be worse. 

I could try the extra layer of inheritance thing, but it doesn't seem clearly better than this to me.

Reply:
Ok the thing I overlooked when i first processed this was [X] which the extra inheritance layer does solve for - so maybe that is the play

Reply:
_with_new & # type: ignore
https://github.com/dagster-io/dagster/pull/22332/commits/473e73917c3cbffea902abec38bbf575b0a35648

vs

add a class layer
https://github.com/dagster-io/dagster/pull/22332/commits/262ec36d4f78f754feb841a641573b20f4370610

Reply:
I prefer the class layer option. Not having `type: ignore` is itself worth it.

Will we need a subclass:
1. anytime we want to define `__new__`? Or,
2. only if the `__new__` types don't match the field types?
I would prefer if we only had one way to override `__new__`, ie option 1 - so the rule is easy to remember. It can also be easily checked I believe in the decorator. But I'm not sure if we need to override `__new__` very often, which can make this quite tedious.

Reply:
Option 3 - can this work as a classmethod factory?

```
@dagster_model
class DbtProject:
    project_dir: Path
    target_path: Path
    target: Optional[str]
    manifest_path: Path
    packaged_project_dir: Optional[Path]
    state_path: Optional[Path]
    has_uninstalled_deps: bool
    manifest_preparer: DbtManifestPreparer

    @classmethod
    def build_from(
        cls,
        project_dir: Union[Path, str],
        *,
        target_path: Union[Path, str] = Path("target"),
        target: Optional[str] = None,
        packaged_project_dir: Optional[Union[Path, str]] = None,
        state_path: Optional[Union[Path, str]] = None,
        manifest_preparer: DbtManifestPreparer = DagsterDbtManifestPreparer(),
    ):
        # coerce logic goes here
        val = cls(
            project_dir=project_dir,
            target_path=target_path,
            target=target,
            manifest_path=manifest_path,
            state_path=project_dir.joinpath(state_path) if state_path else None,
            packaged_project_dir=packaged_project_dir,
            has_uninstalled_deps=has_uninstalled_deps,
            manifest_preparer=manifest_preparer,
        )
        if manifest_preparer:
            manifest_preparer.on_load(val)
        return val

```

Pros:
* doesn't need two classes
* extends to multiple factories, eg `build_from_less_info(...)`, `build_from_more_info(...)`
Cons:
* longer call site expression `DbtProject.build_from(...)`

Reply:
ugh I spoke too soon - the subclassing doesn't work, static analysis doesn't think the child class `__new__` overrides the `@dataclass_transform` based construction 

https://buildkite.com/dagster/dagster-dagster/builds/85351#018feea6-3a3f-4b0e-a928-634fe6033766

Reply:
I think the factory method will work in some cases, but I think its rough for user facing API objects